### DISCUSSION TITLE
Questions about ODE definition

### DISCUSSION THREAD

**Discussion by WayneX-Wong** [idaholab/moose#23698] [Q&A General] [answered]
Hi all,

I am now thinking of a lump parameter problem (or scalar variable problem), of which ODE can be simplified into the following form:
$$A \cdot \frac{d(X \cdot Y)}{dt}=B \cdot F(X)\cdot G(Y)\quad when \quad X>a$$
$$A ^ { \prime } \cdot \frac { d ( X \cdot Y ) } { d t } = B ^ { \prime } \cdot F ^ { \prime } ( X ) \cdot G ^ { \prime } ( Y ) \quad when  \quad a \lt X \lt b$$
There are two variables $X$ and $Y$, and the remaining parameters $A$ and $B$ or the functions $F$ and $G$ are known, and $t$ is time. In order to close this ODE, an equation is added as
$$\xi(Y)=\chi(X)$$

The definition of general time-varying equations in MOOSE should be relatively convenient, directly through the ScalarKernels system. And if the ODE changes with time $t$, it should be realized through the Controls system. However, in the problem I studied, the ODE changes with the variable $X$ to be solved. Does MOOSE have any built-in systems that can meet such requirements?

Thanks!

**Comment by hugary1995** [comment, marked as answer]
Sounds like you are writing your custom scalar kernels anyways, so this should be a if-else condition. Regularize the function domain as you need.

**Reply by WayneX-Wong** [reply]
Thanks! I'll try

**Comment by WayneX-Wong** [comment]
Hi all,

Another question about the ODEs refered above,

$$A \cdot \frac { d ( X \cdot Y ) } { d t } = B \cdot F ( X , Y ) \quad  X \gt a$$

Could rearrange the LHS term (Time deriviate term) into,

$$A \cdot X \cdot \frac{d(Y)}{dt}+A \cdot Y \cdot \frac{d(X)}{dt}$$

So I can define a ODETimekernel about $u$ coupled with $v$ , and use it in the input deck twice wrt $X$ and $Y$ respectively.

While what about RHS? There's no way to separate two variables. Should I define it a $X$ coupled $Y$ or conversely, or both two?

**Reply by hugary1995** [reply]
I didn't quite follow why you want to separate the two variables. Is it for simplicity/modularity?

You should be able to code the exact RHS without any modification. If that doesn't work, let us know why, and we can offer suggestions.

**Reply by WayneX-Wong** [reply]
Thank you for your respond!

Maybe I'm not being explicit. I want to define custom scalarkernel, but RHS have two inseparable variables. But how could a single scalarkernel have two variables? What I'm thinking of is defining an coupled scalarkernel. And that's why I raised the above question.

Please correct if I misunderstood something.

**Reply by hugary1995** [reply]
I see. But you can't solve two unknowns with one equation. You need two equations for two variables.

**Reply by WayneX-Wong** [reply]
@hugary1995 Thank you again for your patience Gary!
Certain it is. In order to close this ODE, another ODE funtion is added in form like:

$$ξ ( Y ) = χ ( X , Y )$$

Which form is similar to the RHS of the ODE above, so I didn‘t mention above, but still have question like this
> Hi all,
> 
> Another question about the ODEs refered above,
> 
> A⋅d(X⋅Y)dt=B⋅F(X,Y)X>a
> 
> Could rearrange the LHS term (Time deriviate term) into,
> 
> A⋅X⋅d(Y)dt+A⋅Y⋅d(X)dt
> 
> So I can define a ODETimekernel about u coupled with v , and use it in the input deck twice wrt X and Y respectively.
> 
> While what about RHS? There's no way to separate two variables. Should I define it a X coupled Y or conversely, or both two?

To be more specific, I wanna refer to the example [Example 3 : Multiphysics coupling](https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex03_coupling.html)
Which has PDEs written as 
$$- \nabla \cdot \nabla u + \nabla v \cdot \nabla u = 0$$
$$- \nabla \cdot \nabla v=0$$
About second term of first PDE, a convection kernel was defined on $u$ with coupled variable $v$. For me it's understandable, cause $v$ can be solved  independently by the second PDE. So when it comes to first PDE, $u$ seems to be more "dominant" than $v$, so $u$ coupled $v$ rather than $v$ coupled $u$ is defined.

But when it comes to my ODEs, variable $X$ and $Y$ have equal status, each variable cannot be determined independently by a single ODE. So I don't know what to do with ODEkernel definitions for terms like $χ ( X , Y )$. Should it be $X$ coupled $Y$ or conversely, or both two?

I’d like to apologize again for the statement that caused me confusion! I am not sure whether my statement is clear this time, and if so, any suggestions would be appreciated very much!

**Reply by hugary1995** [reply]
Okay, no problem. I think I understand now. I guess your confusion is coming from the definition of "variable" in the MOOSE input file. A lot of the times you will see something like this in the input file
```python
[ScalarKernels]
  [foo]
    type = A
    variable = x
    coupled_variable = y
  []
  [bar]
    type = B
    variable = x
  []
  [baz]
    type = C
    variable = y
    coupled_variable = x
  []
[]
```
There isn't any difference between `variable` and `coupled_variable` for the purpose of retrieving values -- their variable values will be made available to you when you compute the residual and Jacobian. The value of the `variable` is retrieved on line 21 in "ScalarKernel.C" as `_u`. The coupled variable value can be retrieved (coupled) on demand using the method `coupledScalarValue`.

There is one important difference, however, for the purpose of assembly. One natural question to ask is: when there are multiple equations and multiple scalar variables, where do their corresponding residuals (and Jacobians) get assembled into the system of equations? The MOOSE convention is that the residual of the scalar kernels are associated with the `variable` of the kernel. In your case, you have two scalar equations in the following form:
```
Equation 1: f(X, Y) = 0
Equation 2: g(X, Y) = 0
```
with two unknowns `X` and `Y`. _You_ need to tell MOOSE how to assemble the residuals. One possibility is to associate the first equation with `X`, and the second equation with `Y`, i.e.
```python
[ScalarKernels]
  [equation1]
    type = Eq1
    variable = X
    coupled_variable = Y
  []
  [equation2]
    type = Eq2
    variable = Y
    coupled_variable = X
  []
[]
```
Another possibility is apparently to switch the `variable` and `coupled_variable` in each kernel. But that's a choice you have to make when you write the source code.

**Reply by WayneX-Wong** [reply]
Okay, I'll try again. In fact, I have already defined according custom scalerkernel.

To be specific

1. $$A \cdot X \cdot \frac { d ( Y ) } { d t } + A \cdot Y \cdot \frac { d ( X ) } { d t }$$
Defined according kernel, time derivative term as `variable` and another as `coupled_variable`, using twice in the input respectively 

1. $$x ( X , Y )$$ $$F ( X , Y )$$
Defined according kernel, using once in the input with $X$ as variable and $Y$ as coupled_variable

But the calculated resudial doesn't deccrease, looks like the equation is not closed, so I have above question.

I will try more, this may take some time, I will update if there is any result.

**Reply by hugary1995** [reply]
It's a system of 2 dofs, so I'd just print out the Jacobian (a 2-by-2 matrix) to verify.

**Reply by WayneX-Wong** [reply]
@hugary1995 My custom kernels runs successfully and get correct result according your guidance
> with two unknowns `X` and `Y`. _You_ need to tell MOOSE how to assemble the residuals. One possibility is to associate the first equation with `X`, and the second equation with `Y`, i.e.

There were other non-convergence problems among the process, which turned out to be due to zero jacobian. I solved the problem by defining zero Jacobian as a small value as [Issue #23594 ](https://github.com/idaholab/moose/discussions/23594) suggests.

Thank you again for your professional guidance and patience Gary!
